File: code-jam/2019-1a/c.cpp
#include <iostream>
#include <map>

using namespace std;

struct tire {
  struct node {
    map<char, node *> children;
    int size;
    node() : size(0) {}
    template <class I> void insert(I begin, I end) {
      size++;
      if (begin == end)
        return;
      if (children.find(*begin) == children.end())
        children[*begin] = new node();
      children[*begin]->insert(begin + 1, end);
    }
    size_t bazinga(bool is_root) const {
      if (size <= 1)
        return 0;
      size_t res = 0;
      for (map<char, node *>::const_iterator i = children.cbegin();
           i != children.cend(); i++)
        res += i->second->bazinga(false);
      res += (!is_root) * 2 * (size - res >= 2);
      return res;
    }
  } root;
  template <class I> void insert(I begin, I end) { root.insert(begin, end); }
  size_t bazinga() { return root.bazinga(true); }
};

void main0() {
  int n;
  tire t;
  cin >> n;
  while (n--) {
    string w;
    cin >> w;
    t.insert(w.rbegin(), w.rend());
  }
  cout << t.bazinga() << '\n';
}

int main() {
  int n_case;
  cin >> n_case;
  for (int i_case = 1; i_case <= n_case; i_case++) {
    cout << "Case #" << i_case << ": ";
    main0();
  }
  return 0;
}

File: code-jam/2019-1a/c.rs
use std::collections::HashMap;
use std::io::stdin;

struct Node {
    children: HashMap<char, Box<Node>>,
    size: usize,
}

impl Node {
    fn new() -> Node {
        Node {
            children: HashMap::new(),
            size: 0,
        }
    }
    fn insert(&mut self, s: &mut Iterator<Item = char>) -> () {
        self.size += 1;
        match s.next() {
            Some(c) => {
                self.children
                    .entry(c)
                    .or_insert(Box::new(Node::new()))
                    .insert(s);
            }
            None => {}
        }
    }
    fn bazinga(&self, is_root: bool) -> usize {
        if self.size <= 1 {
            return 0;
        }
        let mut res = 0;
        for i in self.children.iter() {
            res += i.1.bazinga(false);
        }
        if !is_root && self.size - res >= 2 {
            res += 2;
        }
        return res;
    }
}

struct Tire {
    root: Node,
}

impl Tire {
    fn new() -> Tire {
        Tire { root: Node::new() }
    }
    fn insert(&mut self, s: &mut Iterator<Item = char>) -> () {
        self.root.insert(s);
    }
    fn bazinga(&self) -> usize {
        self.root.bazinga(true)
    }
}

fn next<T>() -> T
where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug,
{
    let mut x: String = "".to_string();
    stdin().read_line(&mut x).unwrap();
    x.trim().parse::<T>().unwrap()
}

fn main0() -> () {
    let mut n: u32 = next();
    let mut t = Tire::new();
    while n != 0 {
        let mut w: String = "".to_string();
        stdin().read_line(&mut w).unwrap();
        t.insert(&mut w.trim().chars().rev());
        n -= 1;
    }
    println!("{}", t.bazinga());
}

fn main() -> () {
    let n_case: u32 = next();
    for i_case in 1..1 + n_case {
        print!("Case #{}: ", i_case);
        main0();
    }
}
